### PR TITLE
WEBUI-104: add functional tests for document import with properties

### PR DESCRIPTION
**Pull request by Pranit-Sotre**
https://jira.nuxeo.com/browse/WEBUI-104

### CODE REVIEW

File: ftest/features/import_doc.feature [hunk 1]
```diff
@@ -37,3 +37,49 @@ Feature: Import and Create Documents
...
+    Then I select an asset type
+    And I add the following properties:
+      | name         | value             |
+      | description  | my description    |
+      | nature       | Application       |
+      | subjects     | Gastronomy,Comics |
+      | expired      | December 28, 2022 |
+    When I click the Create button to complete the import
+    Then I can see that a document of the type Picture and title sample.png is created
+    And I can see the inline nuxeo-image-viewer previewer
```
Comment (Nisha-QA) [inline ftest/features/import_doc.feature:63]: Add a step to validate the added properties

File: ftest/features/import_doc.feature [hunk 2]
```diff
@@ -37,3 +37,49 @@ Feature: Import and Create Documents
...
+      | description  | my description    |
+      | nature       | Application       |
+      | subjects     | Gastronomy,Comics |
+      | expired      | December 28, 2022 |
+    And I click on Apply to all
+    When I click the Create button to complete the import
+    Then I can see 6 documents
+    When I navigate to "sample.png" child
+    Then I can see that a document of the type Picture and title sample.png is created
+    And I can see the inline nuxeo-image-viewer previewer
```
Comment (Nisha-QA) [inline ftest/features/import_doc.feature:98]: Here also, add a step to validate added properties

File: ftest/features/import_doc.feature [hunk 3]
```diff
@@ -37,3 +37,49 @@ Feature: Import and Create Documents
     Then I can see that a document of the type Video and title sample.mp4 is created
     And I can see the inline nuxeo-video-viewer previewer
     And I can see the video conversions panel
+
+  Scenario: Import a Document with adding properties
```
Comment (Nisha-QA) [inline ftest/features/import_doc.feature]: Nit: you can rephrase this to something like : Import a document with properties

**Review by Nisha-QA — COMMENTED** [review]
Since we have many new functions here you can create a new import_dialog.js file and move them

**Review by rakeshkumar1019 — CHANGES_REQUESTED** [review]
make the changes suggested by @Nisha-QA Also, make sure the functional test check passes.

File: ftest/features/import_doc.feature (hunk 3 above)
Comment (Pranit-Sotre) [reply ftest/features/import_doc.feature]: Done changes.

File: ftest/features/import_doc.feature [hunk 4]
```diff
@@ -37,3 +37,59 @@ Feature: Import and Create Documents
...
     And I can see the inline nuxeo-video-viewer previewer
     And I can see the video conversions panel
+
+  Scenario: Import a Document with properties
+    When I click the Create Document button
+    Then I go to the import tab
+    And I can see the import tab content
+    And I upload the sample.png on the tab content page
+    When I click on Add Properties button
+    Then I select an asset type
```
Comment (Gabez0r) [inline ftest/features/import_doc.feature]: A reader won't be able to tell what asset type was selected. This step should actually allow different asset types to be selected (e.g. `I select "Picture" asset type`)

File: packages/nuxeo-web-ui-ftest/pages/ui.js [hunk 5]
```diff
@@ -189,4 +207,24 @@ export default class UI extends BasePage {
...
+
+  fillMultipleValues(table) {
+    table.rows().forEach((row) => {
+      const fieldName = row[0];
+      const fieldEl = this.getField(fieldName);
+      fieldEl.waitForVisible();
+      fieldEl.scrollIntoView();
+      return fixtures.layouts.setValue(fieldEl, row[1]);
+    });
+  }
```
Comment (Gabez0r) [inline packages/nuxeo-web-ui-ftest/pages/ui.js]: Here we are duplicating a bunch of code which is already available in the [document_layout](https://github.com/nuxeo/nuxeo-web-ui/blob/maintenance-3.0.x/packages/nuxeo-web-ui-ftest/pages/ui/browser/document_layout.js#L4-L28) page helper. We should instead have an helper inside the createDialog so that we can retried the (import) document layout and thus reuse the page wrapper methods we already have.

File: packages/nuxeo-web-ui-ftest/pages/ui.js [hunk 6]
```diff
@@ -168,6 +180,12 @@ export default class UI extends BasePage {
     dropdown.click(`#dropdown #contentWrapper div paper-menu div paper-icon-item[name="${selection}"]`);
   }
 
+  asset() {
+    const dropdown = driver.element('nuxeo-select[id="docTypeDropdown"]');
+    dropdown.waitForVisible('paper-item[role="option"]');
+    dropdown.click('paper-item:nth-child(4)');
+  }
+
```
Comment (Gabez0r) [inline packages/nuxeo-web-ui-ftest/pages/ui.js]: This returns an arbitrary item (the 4th item), which makes the method obscure an unpredictable. In case we deploy a new document type, this will might return different results. Also, "asset" as a name does not help much in the context of the ui app. Such an helper would be better placed inside the createDialog.

File: packages/nuxeo-web-ui-ftest/pages/ui.js [hunk 7]
```diff
@@ -79,6 +79,18 @@ export default class UI extends BasePage {
...
+  get selectAnAssetType() {
+    return this.el.element('iron-input input[placeholder="Select an asset type."]');
+  }
+
+  get applyAll() {
+    return this.el.element('div[class="buttons horizontal justified layout"] paper-button.text:nth-child(3)');
+  }
+
```
Comment (Gabez0r) [inline packages/nuxeo-web-ui-ftest/pages/ui.js]: These should all be inside the createDialog (this.ui.createDialog), since we already have an abstraction for this.

File: packages/nuxeo-web-ui-ftest/features/step_definitions/import_dialog.js [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+  this.ui.asset();
+});
+
+Then('I click the Create button to complete the import', function() {
+  const importButton = this.ui.createDialog.importCreateButtonProperties;
+  importButton.waitForVisible();
+  importButton.waitForEnabled();
+  importButton.click();
+  driver.waitForExist('iron-overlay-backdrop', driver.options.waitForTimeout, true);
+  driver.pause(3000); // XXX just give it some time to the server to do the conversions
```
Comment (Gabez0r) [inline packages/nuxeo-web-ui-ftest/features/step_definitions/import_dialog.js]: This will render the test flaky. If you need to wait for longer, you can always wait for the toast message.

File: packages/nuxeo-web-ui-ftest/features/step_definitions/import_dialog.js [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+});
+
+Then('I can see the following properties:', function(table) {
```
Comment (Gabez0r) [inline packages/nuxeo-web-ui-ftest/features/step_definitions/import_dialog.js]: Why do we need to create a new step and don't reuse the other step we already use to check properties after creating documents? https://github.com/nuxeo/nuxeo-web-ui/blob/maintenance-3.0.x/ftest/features/create_doc.feature#L51

File: packages/nuxeo-web-ui-ftest/features/step_definitions/import_dialog.js [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+  });
+});
+
+Then('I add the following properties:', function(table) {
+  this.ui.fillMultipleValues(table);
+});
+
+Then('I click on Apply to all', function() {
+  this.ui.applyAll.click();
+});
```
Comment (Gabez0r) [inline packages/nuxeo-web-ui-ftest/features/step_definitions/import_dialog.js]: This should live inside the createDialog (this.ui.createDialog).

**Review by Gabez0r — CHANGES_REQUESTED** [review]
It's going in the right direction, good job!

File: packages/nuxeo-web-ui-ftest/pages/ui.js (hunk 5 above)
Comment (Pranit-Sotre) [reply packages/nuxeo-web-ui-ftest/pages/ui.js]: We are not using the existing function, as xpath is different for create and import dialog box.

File: packages/nuxeo-web-ui-ftest/features/step_definitions/import_dialog.js (hunk 9 above)
Comment (Pranit-Sotre) [reply packages/nuxeo-web-ui-ftest/features/step_definitions/import_dialog.js]: We need a new step as web element location is different when we import a document.

**Comment by richardsd** [comment]
@Pranit-Sotre don't forget to rebase your branch. I see the `ftests` pr check failing on the install step (it could help).

File: packages/nuxeo-web-ui-ftest/features/step_definitions/import_dialog.js (hunk 9 above)
Comment (Gabez0r) [reply packages/nuxeo-web-ui-ftest/features/step_definitions/import_dialog.js]: This step validates the metadata after navigating to the document that was just created, so I would not expect the element location to be different.

File: packages/nuxeo-web-ui-ftest/pages/ui.js (hunk 5 above)
Comment (Gabez0r) [reply packages/nuxeo-web-ui-ftest/pages/ui.js]: I'm not following. You should have a helper that returns the a [DocumentLayout](https://github.com/nuxeo/nuxeo-web-ui/blob/maintenance-3.0.x/packages/nuxeo-web-ui-ftest/pages/ui/browser/document_layout.js#L4-L28) wrapping your import layout, which should allow you to reuse the helpers.

File: ftest/features/import_doc.feature [hunk 11]
```diff
@@ -37,3 +37,59 @@ Feature: Import and Create Documents
...
+    Then I can see 6 documents
+    When I navigate to "sample.png" child
+    Then I can see that a document of the type Picture and title sample.png is created
+    And I can see the following properties:
+      | name         | value             |
+      | title        | sample.png        |
+      | description  | my description    |
+      | nature       | Application       |
+    And I can see the inline nuxeo-image-viewer previewer
+    And I can see the picture formats panel
```
Comment (Gabez0r) [inline ftest/features/import_doc.feature]: nit: Missing newline at the end of the file

File: packages/nuxeo-web-ui-ftest/pages/ui/create_dialog.js [hunk 12]
```diff
@@ -58,6 +58,10 @@ export default class CreateDialog extends BasePage {
     return this.el.element('div[name="upload"] paper-button[id="create"]');
   }
 
+  get importCreateButtonProperties() {
+    return this.el.element('div[class="buttons horizontal justified layout"] paper-button.primary');
```
Comment (Gabez0r) [inline packages/nuxeo-web-ui-ftest/pages/ui/create_dialog.js]: We should probably add a `name`, otherwise this will break if someone adjusts a class. for example.

File: packages/nuxeo-web-ui-ftest/pages/ui/create_dialog.js [hunk 13]
```diff
@@ -85,4 +89,24 @@ export default class CreateDialog extends BasePage {
   get createButton() {
     return this.el.element('paper-button[id="create"]');
   }
+
+  get addProperties() {
+    return this.el.element('paper-button[id="edit"]');
+  }
+
+  get selectAnAssetType() {
+    return this.el.element('iron-input input[placeholder="Select an asset type."]');
```
Comment (Gabez0r) [inline packages/nuxeo-web-ui-ftest/pages/ui/create_dialog.js]: Same as above, adding a name or id here is preferable, since the placeholder name might change.

File: packages/nuxeo-web-ui-ftest/pages/ui/create_dialog.js [hunk 14]
```diff
@@ -85,4 +89,24 @@ export default class CreateDialog extends BasePage {
...
+  }
+
+  get applyAll() {
+    return this.el.element('div[class="buttons horizontal justified layout"] paper-button.text:nth-child(3)');
```
Comment (Gabez0r) [inline packages/nuxeo-web-ui-ftest/pages/ui/create_dialog.js]: Same as above, adding a name here is preferable.

File: packages/nuxeo-web-ui-ftest/pages/ui/create_dialog.js [hunk 15]
```diff
@@ -85,4 +89,24 @@ export default class CreateDialog extends BasePage {
...
+  }
+
+  asset(val) {
```
Comment (Gabez0r) [inline packages/nuxeo-web-ui-ftest/pages/ui/create_dialog.js]: This method should be named after what it does. For example, "selectAsset".

File: elements/document/nuxeo-document-import.js [hunk 16]
```diff
@@ -678,6 +680,7 @@ Polymer({
           <paper-button
             noink
             class="primary"
+            name="create"
```
Comment (Gabez0r) [inline elements/document/nuxeo-document-import.js]: There is already another button that does creation (with id="create"), so perhaps we could make this name more explicit to avoid any confusions. For example "createWithProperties" or something that tells that this is going to be used in the advanced scenario where we are adding properties.

File: packages/nuxeo-web-ui-ftest/pages/ui/create_dialog.js [hunk 17]
```diff
@@ -1,12 +1,17 @@
...
 
 export default class CreateDialog extends BasePage {
   get documentCreate() {
     return new DocumentCreate('nuxeo-document-create');
   }
 
+  get documentLayout() {
+    return new DocumentLayout('nuxeo-document-import');
+  }
+
```
Comment (Gabez0r) [inline packages/nuxeo-web-ui-ftest/pages/ui/create_dialog.js]: This is confusing because from this element, we can reach both "create" and "import" layouts. We have two options here:
1) Looking at the lines immediately above, I believe we could further extract this into a `DocumentImport` wrapper, which then could return the layout just have as you have here. This would also be consistent to what we are doing with DocumentCreate. This would allow us to move all the helpers we have below to this new wrapper.
2) Simpler option, but probably providing a more disperse API, to rename this method to "documentImportLayout".

**Review by Gabez0r — COMMENTED** [review]
Looking good. We are almost there!

**Review by Nisha-QA — DISMISSED** [review]
Looks good

**Review by rajatgoel-gl — DISMISSED** [review]

File: packages/nuxeo-web-ui-ftest/pages/ui/create_dialog.js [hunk 18]
```diff
@@ -85,4 +94,24 @@ export default class CreateDialog extends BasePage {
...
+  get selectAnAssetType() {
+    return this.el.element('nuxeo-select[name="assetType"]');
+  }
+
+  get applyAll() {
+    return this.el.element('paper-button[name="applyAll"]');
+  }
+
+  selectAsset(val) {
```
Comment (Gabez0r) [inline packages/nuxeo-web-ui-ftest/pages/ui/create_dialog.js]: I believe `selectAssetType` would be more appropriate even.

File: packages/nuxeo-web-ui-ftest/features/step_definitions/import_dialog.js [hunk 19]
```diff
@@ -0,0 +1,28 @@
...
+  this.ui.createDialog.selectAnAssetType.click();
+  this.ui.createDialog.selectAsset(val);
+});
+
+Then('I click the Create button to complete the import', function() {
+  const importButton = this.ui.createDialog.importCreateButtonProperties;
+  importButton.waitForVisible();
+  importButton.waitForEnabled();
+  importButton.click();
+  driver.waitForExist('#snackbarPanel mwc-snackbar[open] .mdc-snackbar__label', driver.options.waitForTimeout, true);
```
Comment (Gabez0r) [inline packages/nuxeo-web-ui-ftest/features/step_definitions/import_dialog.js]: This just waits for the toast to be dismissed, but we are not even sure what message it was displaying, so this step might become flaky in some situations. We should wait to see the proper message on the toast that is displayed once the creation is complete.

**Review by rakeshkumar1019 — COMMENTED** [review]
Squash the last commit, it looks good.

File: packages/nuxeo-web-ui-ftest/features/step_definitions/import_dialog.js (hunk 19 above)
Comment (richardsd) [reply packages/nuxeo-web-ui-ftest/features/step_definitions/import_dialog.js]: I agree with @Gabez0r. We could probably do something like the following to assert that the message is displayed:

```js
  const notificationMessage = this.ui.getToastMessage(message);
  const trimmedMessage = message.trim().replace(/"/g, '');
  notificationMessage.should.be.equals(trimmedMessage);
```
Check https://github.com/nuxeo/nuxeo-web-ui/blob/maintenance-3.0.x/packages/nuxeo-web-ui-ftest/features/step_definitions/bulk_edit.js#L14-L18 for the context.

**Review by richardsd — CHANGES_REQUESTED** [review]
Please review your code based on @Gabez0r comment 🙂

File: packages/nuxeo-web-ui-ftest/features/step_definitions/import_dialog.js (hunk 19 above)
Comment (Pranit-Sotre) [reply packages/nuxeo-web-ui-ftest/features/step_definitions/import_dialog.js]: I have covered this point in import_doc.feature at line number 55 and 88.

Comment (Gabez0r) [reply packages/nuxeo-web-ui-ftest/features/step_definitions/import_dialog.js]: Ok then this this line waiting for the  toast to disappear is redundant.

**Review by FabioFSantos — CHANGES_REQUESTED** [review]
Please review your code based on @Gabez0r last comment about the toast check redundancy, the rest looks good

File: packages/nuxeo-web-ui-ftest/features/step_definitions/import_dialog.js (hunk 19 above)
Comment (Pranit-Sotre) [reply packages/nuxeo-web-ui-ftest/features/step_definitions/import_dialog.js]: Done

**Review by richardsd — DISMISSED** [review]

**Review by Nisha-QA — DISMISSED** [review]
Looks good

**Review by FabioFSantos — DISMISSED** [review]

**Review by rakeshkumar1019 — APPROVED** [review]
looks good 👍

**Review by richardsd — APPROVED** [review]

**Review by Gabez0r — APPROVED** [review]

**Review by FabioFSantos — APPROVED** [review]

**Review by pddgoncalves — APPROVED** [review]

**Review by Nisha-QA — APPROVED** [review]
Looks good
